pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go

```diff
@@ -368,8 +461,9 @@ func (plan *TestPlan) Steps() []testStep {
 // Versions returns the list of versions the cluster goes through in
 // the process of running this mixed-version test.
 func (plan *TestPlan) Versions() []*clusterupgrade.Version {
-	result := []*clusterupgrade.Version{plan.upgrades[0].from}
-	for _, upgrade := range plan.upgrades {
+	upgrades := plan.allUpgrades()
+	result := []*clusterupgrade.Version{upgrades[0].from}
```

Comment:
Not immediately clear if this list may contain duplicates?

Reply:
There are no duplicates (if there were, that would be an internal bug in the planner).

`allUpgrades` returns a list of all setup upgrades, followed by the test's upgrades. These run in order. An upgrade can either be run during setup or during "test" time, along with user functions.

I added a note about this in the comments, hopefully it makes this clearer.